torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -1693,7 +1693,15 @@ int dictLen(Stack& stack) {
 
 template <unsigned int Index, typename Elem>
 c10::List<Elem> makeListForDictKeysOrValues(
+    const std::pair<c10::optional<TypePtr>, c10::optional<TypePtr>>& types,
     const std::vector<std::pair<IValue, IValue>>& order) {
+  TORCH_INTERNAL_ASSERT(
```

Comment:
What is the point of the changes to this function ? it already worked

Reply:
Additional error checking that the lists we're passing around have the correct type

Reply:
the typesystem should already ensure that it's the correct type. if we're adding runtime enforcements here shouldn't we be adding a runtime enforcement to every op that involves a list?

Reply:
It's not a huge deal either way but I still would prefer not to set the precedent that we don't trust the input types of the statically typed langauge. 

@zdevito thoughts ?

Reply:
to reiterate our offline discussion: What this code **should** do is take the type from the source dict and write them to the target list, same as what `makeGenericListForDictKeysOrValues` is doing.

Unfortunately, only `GenericList` has a constructor that takes a runtime type and can do this. `List<T>` doesn't need that because it infers the runtime type from it's `T` type parameter. So instead of passing in the runtime type to the constructor, this is an assertion to check that the inferred type is the one we expect to be and would have passed in if we had created a `GenericList`.

Note that this is a `TORCH_INTERNAL_ASSERT` and not a `TORCH_CHECK` because this is not a type check that reports errors to users. Static typing should make sure that this never fails. But it can fail if there's a PyTorch bug and I'd rather we catch that here than just ignore it.

Reply:
This is just an assert, so it is ok. It improves the error message that would otherwise happen in the `to` call.